packages/database/src/tests/cachers/EntityHierarchyCacher.test.js

```diff
@@ -36,23 +54,171 @@ describe('EntityHierarchyCacher', () => {
+  });
+
+  it('deletes and rebuilds a subtree if entity relation records are changed', async () => {
+    // change ab to sit below aab instead of aa
+    await models.entityRelation.update(
+      {
+        parent_id: 'entity_aa_test',
+        child_id: 'entity_ab_test',
+        entity_hierarchy_id: 'hierarchy_b_test',
+      },
+      { parent_id: 'entity_aab_test' },
+    );
+    // change aab to sit below aa instead of abb
+    await models.entityRelation.update(
+      {
+        parent_id: 'entity_abb_test',
+        child_id: 'entity_aab_test',
+        entity_hierarchy_id: 'hierarchy_b_test',
+      },
+      { parent_id: 'entity_aa_test' },
+    );
+    await assertRelationsMatch('project_b_test', HIERARCHY_B_AFTER_MULTIPLE_RELATIONS_CHANGED);
+  });
+
+  it('adds a new subtree if entity relation records are added', async () => {
+    // change ab to sit below aab instead of aa
+    await models.entityRelation.create({
+      parent_id: 'entity_ab_test',
+      child_id: 'entity_aba_test',
+      entity_hierarchy_id: 'hierarchy_b_test',
+    });
+    await assertRelationsMatch('project_b_test', HIERARCHY_B_AFTER_ENTITY_RELATION_ADDED);
+  });
+
+  it('adds new subtrees if an entities are created', async () => {
+    // update the parent_id of an entity, and make sure the subtree in the database is rebuilt
+    await upsertDummyRecord(models.entity, { id: 'entity_ac_test', parent_id: 'entity_a_test' });
+    await upsertDummyRecord(models.entity, { id: 'entity_aca_test', parent_id: 'entity_ac_test' });
+    await upsertDummyRecord(models.entity, { id: 'entity_abc_test', parent_id: 'entity_ab_test' });
+    await upsertDummyRecord(models.entity, {
+      id: 'entity_aaaa_test',
+      parent_id: 'entity_aaa_test',
     });
+    await assertRelationsMatch('project_a_test', HIERARCHY_A_AFTER_ENTITIES_CREATED);
+    await assertRelationsMatch('project_b_test', HIERARCHY_B_AFTER_ENTITIES_CREATED);
+  });
+
+  it('batches multiple changes', async () => {
+    sinon.stub(EntityHierarchyCacher.prototype, 'buildAndCacheHierarchies');
+
+    // make a bunch of different changes, with small delays between each to model real life
+    const sleepTime = 250; // sleep for 250 ms between each change
+
+    // create an entity
+    await upsertDummyRecord(models.entity, { id: 'entity_ac_test', parent_id: 'entity_a_test' });
+    await sleep(sleepTime);
+
+    // update an entity relation
+    await models.entityRelation.update(
+      {
+        parent_id: 'entity_aa_test',
+        child_id: 'entity_ab_test',
+        entity_hierarchy_id: 'hierarchy_b_test',
+      },
+      { parent_id: 'entity_aab_test' },
+    );
+    await sleep(sleepTime);
+
+    // delete an entity
+    const entityToDelete = 'entity_aaa_test';
+    await models.entityRelation.delete({ child_id: entityToDelete }); // can't delete entity until entity relation is gone
+    await models.entity.delete({ id: entityToDelete });
```

Comment:
Yeah it's out of scope, but I'm also not too mad at leaving it as is. A developer can always delete the entity relations as a response to an entity deletion request from whatever client is asking, but it also provides a double check that we really intend to delete this entity, even though it may be involved in hierarchies other than the canonical one. Still, you're probably right, but yes, best deferred to another task.